### ISSUE TITLE
CMake configuration overhaul

### ISSUE THREAD

**Issue by tmadlener** [iLCSoft/LCIO#133]
The cmake configuration of several ilcsoft packages is pretty old and could potentially profit from exploiting newer cmake functionalities and in general newer cmake paradigms, e.g. target based configurations. Since LCIO is at the very base of everything, it could serve as a good starting point to check what would need changing and could then serve as an example for other packages as well.

A few points that should be decided/considered beforehand:
- [ ] Which minimum cmake version do we require? LCG releases 96 and 97 come with 3.14, which looks like it is probably the hardest constraint. The current stable version is 3.20, Ubuntu 20.04 has 3.16
- [ ] "Deprecation strategy" for the old configuration. In principle the target based and `XXX_LIBRARIES`, `XXX_INCLUDE_DIRS` approach can co-exist (see, e.g. [SIO](https://github.com/iLCSoft/SIO/blob/master/cmake/SIOConfig.cmake.in), so that it should be possible to update this package by package without breaking everything while updating.

A more modern cmake configuration has the potential to make some of the cmake macros that are part of ilcsoft obsolete, as cmake can now handle many things for which these macros were originally designed by itself. To which degree we can remove these macros is hard to estimate and will be something we find out along the way.

**Comment by tmadlener** [comment]
Arguments for going to cmake > 3.15: https://cmake.org/cmake/help/latest/policy/CMP0094.html#policy:CMP0094

**Comment by tmadlener** [comment]
https://github.com/iLCSoft/LCIO/pull/132#issuecomment-832071129 for two more points:

- Use `FindPython` to detect python
- Make sure the `BUILD_TESTING` option is considered for all tests

**Comment by tmadlener** [comment]
Some more considerations for the choice of the minimum required cmake version:
https://github.com/iLCSoft/LCIO/pull/132#issuecomment-832555662

**Comment by Romendakil** [comment]
This remark fits to this issue, but feel free to open a separate one: I noticed that the two files LCIOConfig.cmake and LCIOConfigVersion.cmake end up after 'make install' in <CMAKE_INSTALL_PREFIX> directly, while for most (maybe all?) cmake built projects I know they are installed/written to <CMAKE_INSTALL_PREFIX>/lib/cmake
I would find it more consistent for these files to end up of there.

**Comment by tmadlener** [comment]
Yes, definitely part of this issue. I need to find some time to finalize #140 and then this should also be fixed.

**Comment by jan-busa** [comment]
When I build latest version v2-20-2, ```LCIOConfig.cmake``` is placed under ```lib/cmake/LCIO/LCIOConfig.cmake``` and ```SIOTargets.cmake``` goes into ```lib/cmake/SIO/SIOTargets.cmake``` (notice the different directories). This leads to the problem when ```LCIOConfig.cmake``` around line 75 looks for ```SIOTargets.cmake``` in the line ```include("${CMAKE_CURRENT_LIST_DIR}/SIOTargets.cmake")```. Of course the easiest way for fix would be to hard-code SIO path as ```include("${CMAKE_CURRENT_LIST_DIR}/.../SIO//SIOTargets.cmake")``` but I'm sure, there is a more elegant way, I'm just not so familiar with CMake.

**Comment by tmadlener** [comment]
I have moved your issue (@jan-busa) into a separate one, because I this one is much broader and more work in general than solving yours (I hope)
